sql: implement FILTER clauses for aggregate expressions

andreimatei opened the pull request:
This is a nifty feature introduced in SQL:2003 and supported by Postgres
since 9.4.
If FILTER is specified in an aggregation expression, then only the
input rows for which the filter clause evaluates to true are fed to the
aggregate function; other rows are discarded. For example:

SELECT
    count(_) AS unfiltered,
    count(_) FILTER (WHERE i < 5) AS filtered
    FROM generate_series(1,10) AS s(i);
unfiltered | filtered
-----------+----------
10         | 4
(1 row)

There's still a problem with this implementation: the grammar
(copied from Postgres) accepts FILTER clauses on function calls that are
not aggregations. I need to figure out how to trap that - suggestions
welcomed!
Postgres does this:
test=# select abs(a) filter (where a>0) from t;
ERROR:  FILTER specified, but abs is not an aggregate function

test=# select i, rank() filter (where marker=true) over (order by i),
max(i) filter (where marker=true) over(order by i) from a;
ERROR:  FILTER is not implemented for non-aggregate window functions

Next I'll implement support for these clauses in window functions.

CC @knz

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10146)

<!-- Reviewable:end -->

andreimatei committed:
sql: implement FILTER clauses for aggregate expressions

This is a nifty feature introduced in SQL:2003 and supported by Postgres
since 9.4.
If FILTER is specified in an aggregation expression, then only the
input rows for which the filter clause evaluates to true are fed to the
aggregate function; other rows are discarded. For example:

SELECT
    count(*) AS unfiltered,
    count(*) FILTER (WHERE i < 5) AS filtered
    FROM generate_series(1,10) AS s(i);
unfiltered | filtered
-----------+----------
10         | 4
(1 row)

There's still a problem with this implementation: the grammar
(copied from Postgres) accepts FILTER clauses on function calls that are
not aggregations. I need to figure out how to trap that - suggestions
welcomed!
Postgres does this:
test=# select abs(a) filter (where a>0) from t;
ERROR:  FILTER specified, but abs is not an aggregate function

test=# select i, rank() filter (where marker=true) over (order by i),
max(i) filter (where marker=true) over(order by i) from a;
ERROR:  FILTER is not implemented for non-aggregate window functions

Next I'll implement support for these clauses in window functions.

knz commented:
To check whether FILTER is properly applied and reject invalid uses, modify FuncExpr.TypeCheck() in parser/type_check.go.

---

Review status: 0 of 8 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/sql/group.go, line 96 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-KUfZjwTufVfFbTeqRmy:-KUfZjwVRCdt5XmPunXJ:bkdv08t) ([raw file](https://github.com/cockroachdb/cockroach/blob/c55c2e79286f9596bb6969bf7a193a5248af8204/pkg/sql/group.go#L96)):*
> ```Go
> 		values:             valuesNode{columns: s.columns},
> 		render:             s.render,
> 		filterToRenderIdxs: make(map[int]int),
> ```

I retract my argument fro,  yesterday on slack: this is probably the right way to deal with grouping filters. Thanks for figuring this out.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-:-KUf_Cg-RPzXaFO-oS4r:b-ronzaq)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Reviewed 7 of 8 files at r1.
Review status: 7 of 8 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/sql/group.go, line 343 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-KUm3_0F65_LIPHS7_Qm:-KUm3_0F65_LIPHS7_Qn:b3rvhiy) ([raw file](https://github.com/cockroachdb/cockroach/blob/c55c2e79286f9596bb6969bf7a193a5248af8204/pkg/sql/group.go#L343)):*
> ```Go
> 				} else {
> 					filterVal := values[renderIdx]
> 					if filterVal.Compare(parser.DBoolTrue) != 0 {
> ```

Does this handle Nulls correctly? Check out`sqlbase.RunFilter` for similar logic.

---

*[pkg/sql/group.go, line 559 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-KUm5jGKFh8JuDmo8_FQ:-KUm5jGKFh8JuDmo8_FR:b-8umg8s) ([raw file](https://github.com/cockroachdb/cockroach/blob/c55c2e79286f9596bb6969bf7a193a5248af8204/pkg/sql/group.go#L559)):*
> ```Go
> 			var filterExpr parser.TypedExpr
> 			if t.Filter != nil {
> 				filterExpr = t.Filter.(parser.TypedExpr)
> ```

It doesn't look like `t.Filter` is ever type checked. I think you're missing logic in `parser/type_check.go` and also `parser/walk.go`.

---

*[pkg/sql/logic_test.go, line 1002 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-KUm1_azUeek2Q5DhTSm:-KUm1_b-BV025pF3RAfk:bd0vbvh) ([raw file](https://github.com/cockroachdb/cockroach/blob/c55c2e79286f9596bb6969bf7a193a5248af8204/pkg/sql/logic_test.go#L1002)):*
> ```Go
> 			fmt.Fprintf(tw, "    %s\n", expectedResultRaw)
> 		}
> 		// TODO(andrei): Figure out how to present the results instead of
> ```

Interesting, I've never noticed this before. So the issue is only that we display the rows in the incorrect sorting order if the (correctly sorted) comparison fails, not that we compare the rows in the incorrect sorting order, correct?

---

*[pkg/sql/window.go, line 723 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-KUm27GEQ-EDcFZ1HzBp:-KUm27GEQ-EDcFZ1HzBq:blh0yeu) ([raw file](https://github.com/cockroachdb/cockroach/blob/c55c2e79286f9596bb6969bf7a193a5248af8204/pkg/sql/window.go#L723)):*
> ```Go
> 			}
> 
> 			if t.Filter != nil {
> ```

Like @knz mentioned above, this kind of logic can been handled in `parser/type_check.go`. Specifically, it can fit into [this logic in FuncExpr.TypeCheck](https://github.com/cockroachdb/cockroach/blob/a3e6319c27ebf07be4c048baac2c6bd1a785d1b3/pkg/sql/parser/type_check.go#L318). You can also add corresponding tests to `TestTypeCheckError`.

---

*[pkg/sql/parser/expr.go, line 806 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-KUm1EmxHRhQMPjk81iq:-KUm1EmxHRhQMPjk81ir:b7bfa63) ([raw file](https://github.com/cockroachdb/cockroach/blob/c55c2e79286f9596bb6969bf7a193a5248af8204/pkg/sql/parser/expr.go#L806)):*
> ```Go
> 	Name      NormalizableFunctionName
> 	Type      funcType
> 	Filter    Expr
> ```

Similar nit to below: swap `Filter` and `Exprs`

---

*[pkg/sql/parser/sql.y, line 3718 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-KUm0tFnYXRQeDM9mPVV:-KUm0tFnYXRQeDM9mPVW:bwe4e5u) ([raw file](https://github.com/cockroachdb/cockroach/blob/c55c2e79286f9596bb6969bf7a193a5248af8204/pkg/sql/parser/sql.y#L3718)):*
> ```Yacc
>   {
>     f := $1.expr().(*FuncExpr)
>     f.WindowDef = $4.windowDef()
> ```

nit: Let's order these in the order we see the variables (`$1, $3, $4`)

---

*[pkg/sql/testdata/aggregate, line 773 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-KUm0IwSD4vQYtdDP3iy:-KUm0IwTYJluq9KF5U4_:b-1wd39n) ([raw file](https://github.com/cockroachdb/cockroach/blob/c55c2e79286f9596bb6969bf7a193a5248af8204/pkg/sql/testdata/aggregate#L773)):*
> ```
> 
> statement ok
> DROP TABLE kv
> ```

It would be preferable to either use the existing kv table or add a new table, so that the contents of `kv` doesn't depend on where in the file we are.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 7 of 8 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/sql/testdata/aggregate, line 808 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-KUmu_GNJQ7FsA_9AfZo:-KUmu_GNJQ7FsA_9AfZp:b-h69mhr) ([raw file](https://github.com/cockroachdb/cockroach/blob/c55c2e79286f9596bb6969bf7a193a5248af8204/pkg/sql/testdata/aggregate#L808)):*
> ```
> 4 true 1 2 3
> NULL true 0 1 5
> ```

This could also use some tests with different `EXPLAIN` modes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Cool stuff! We should add a couple of testcases for the new syntax in TestParse.

RaduBerinde commented:
CC @irfansharif in case you want to think about how we could go about adding this to distsql aggregation.

andreimatei commented:
@knz moved type checking to the right place, thanks

---

Review status: 0 of 11 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[pkg/sql/group.go, line 96 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-KUfZjwTufVfFbTeqRmy:-K_8B1XQgleGLWy6Y7-D:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c55c2e79286f9596bb6969bf7a193a5248af8204/pkg/sql/group.go#L96)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I retract my argument fro,  yesterday on slack: this is probably the right way to deal with grouping filters. Thanks for figuring this out.
</blockquote></details>

Done.

---

*[pkg/sql/group.go, line 343 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-KUm3_0F65_LIPHS7_Qm:-K_813T9s4cOjKjx9BE6:b-tw6nrq) ([raw file](https://github.com/cockroachdb/cockroach/blob/c55c2e79286f9596bb6969bf7a193a5248af8204/pkg/sql/group.go#L343)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Does this handle Nulls correctly? Check out`sqlbase.RunFilter` for similar logic.
</blockquote></details>

I believe it was correct, but switched to what `RunFilter` is doing. I guess `datum.Compare` deals with ordering, and it's less intuitive how the nulls will behave.

---

*[pkg/sql/group.go, line 559 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-KUm5jGKFh8JuDmo8_FQ:-K_7MvXNA7P0VPaAUwK3:b1i3dhr) ([raw file](https://github.com/cockroachdb/cockroach/blob/c55c2e79286f9596bb6969bf7a193a5248af8204/pkg/sql/group.go#L559)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

It doesn't look like `t.Filter` is ever type checked. I think you're missing logic in `parser/type_check.go` and also `parser/walk.go`.
</blockquote></details>

good catch, added

---

*[pkg/sql/group.go, line 208 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_8BgnVBwGFh7i8ehFa:-K_8BgnVBwGFh7i8ehFb:bydz03h) ([raw file](https://github.com/cockroachdb/cockroach/blob/428f525528d39721afede723905ad7e14008cae7/pkg/sql/group.go#L208)):*
> ```Go
> 		// aggregate function so the merge loop cannot encounter an
> 		// equivalent expression in the aggregation list.
> 		_ = s.addOrMergeRenders(cols, exprs, false)
> ```

I don't quite understand what this comment is trying to tell me. Given that the `reuseExistingRender` arg is `false`, does the fact that the same expression cannot be used as both a group by and an output matter?
Could we pass `reuseExistingRender = true` here and/or below, in the new all to `addOrMergeRenders`? I guess I don't understand when renders cannot be reused.

---

*[pkg/sql/logic_test.go, line 1002 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-KUm1_azUeek2Q5DhTSm:-K_7QEleIgU2Qbx_dJ5U:bzbglpd) ([raw file](https://github.com/cockroachdb/cockroach/blob/c55c2e79286f9596bb6969bf7a193a5248af8204/pkg/sql/logic_test.go#L1002)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Interesting, I've never noticed this before. So the issue is only that we display the rows in the incorrect sorting order if the (correctly sorted) comparison fails, not that we compare the rows in the incorrect sorting order, correct?
</blockquote></details>

correct.
@knz, I remember you touching and fixing related stuff recently. Does this ring a bell?

---

*[pkg/sql/parser/expr.go, line 806 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-KUm1EmxHRhQMPjk81iq:-K_81k4AzlfXo3xgotMa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c55c2e79286f9596bb6969bf7a193a5248af8204/pkg/sql/parser/expr.go#L806)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Similar nit to below: swap `Filter` and `Exprs`
</blockquote></details>

Done.

---

*[pkg/sql/parser/sql.y, line 3718 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-KUm0tFnYXRQeDM9mPVV:-K_81_EYjZ3-zIjoBDmC:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c55c2e79286f9596bb6969bf7a193a5248af8204/pkg/sql/parser/sql.y#L3718)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: Let's order these in the order we see the variables (`$1, $3, $4`)
</blockquote></details>

done

---

*[pkg/sql/testdata/aggregate, line 773 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-KUm0IwSD4vQYtdDP3iy:-K_89MBGuZuLbREI61aa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c55c2e79286f9596bb6969bf7a193a5248af8204/pkg/sql/testdata/aggregate#L773)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

It would be preferable to either use the existing kv table or add a new table, so that the contents of `kv` doesn't depend on where in the file we are.
</blockquote></details>

Done.

---

*[pkg/sql/testdata/aggregate, line 808 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-KUmu_GNJQ7FsA_9AfZo:-K_89MnDHj_SvulTwIH9:bqlyccn) ([raw file](https://github.com/cockroachdb/cockroach/blob/c55c2e79286f9596bb6969bf7a193a5248af8204/pkg/sql/testdata/aggregate#L808)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This could also use some tests with different `EXPLAIN` modes.
</blockquote></details>

well I've added an `EXPLAIN(TYPES)` and I think we're seeing what we want, but someone double-checking would be great

---

*[pkg/sql/window.go, line 723 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-KUm27GEQ-EDcFZ1HzBp:-K_7O7YaBLX0AdFFa3i0:bhgcyrr) ([raw file](https://github.com/cockroachdb/cockroach/blob/c55c2e79286f9596bb6969bf7a193a5248af8204/pkg/sql/window.go#L723)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Like @knz mentioned above, this kind of logic can been handled in `parser/type_check.go`. Specifically, it can fit into [this logic in FuncExpr.TypeCheck](https://github.com/cockroachdb/cockroach/blob/a3e6319c27ebf07be4c048baac2c6bd1a785d1b3/pkg/sql/parser/type_check.go#L318). You can also add corresponding tests to `TestTypeCheckError`.
</blockquote></details>

excellent, done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-:-K_7OCQvoO3A5mrm0u5D:b-eyu2mb)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 10 of 12 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[pkg/sql/group.go, line 208 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_8BgnVBwGFh7i8ehFa:-K_A16MOHeNn2kR_T7zw:bu2mpkn) ([raw file](https://github.com/cockroachdb/cockroach/blob/428f525528d39721afede723905ad7e14008cae7/pkg/sql/group.go#L208)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I don't quite understand what this comment is trying to tell me. Given that the `reuseExistingRender` arg is `false`, does the fact that the same expression cannot be used as both a group by and an output matter?
Could we pass `reuseExistingRender = true` here and/or below, in the new all to `addOrMergeRenders`? I guess I don't understand when renders cannot be reused.
</blockquote></details>

This comment is obviously garbage since the argument is false. I do not know why I forgot to remove it when I committed 9450059ada3d31fe73f0fe4573f4b39b4cbb9d79. 
FWIW I just made a small experiment by setting the argument to `true` and it breaks badly. I do not know yet why it breaks, but that is why the argument is currently `false` and probably why the comment was written (I probably wrote the comment first, thinking it could be `true`, then tweaked the argument later without realizing the comment was stale.)

Note that yes, we'll want to merge renders eventually. This is definitely on my radar still (here and in other places). However we need top clean up some other code before we can solve this issue.

---

*[pkg/sql/group.go, line 233 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_A1sKrC475UDrxdX_O:-K_A1sKs52FeZDiDunlU:b-ldli8p) ([raw file](https://github.com/cockroachdb/cockroach/blob/922c218ebbfd70e5f8d2e761e704c6709612ac52/pkg/sql/group.go#L233)):*
> ```Go
> 			panic("star found in filter; this should not have passed type checking")
> 		}
> 		colIdxs := s.addOrMergeRenders(cols, exprs, false /* reuseExistingRender */)
> ```

In this specific case I think you can request the merge with `true` and it should work, because you track the mapping manually in `filterToRenderIdxs`. Have you tried? Ideally with this if you have `select count(*) filter(where x+1=3), max(x) filter(where x+1=3) from ...` you should see `x+1=3` rendered just once (use EXPLAIN(EXPRS) to check).

---

*[pkg/sql/group.go, line 398 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_A2kPX6Q3Ym9eFVmod:-K_A2kPYNGexIy3gKmsa:b1igglm) ([raw file](https://github.com/cockroachdb/cockroach/blob/922c218ebbfd70e5f8d2e761e704c6709612ac52/pkg/sql/group.go#L398)):*
> ```Go
> 
> 		// Feed the aggregateFuncHolders for this bucket the non-grouped values.
> 		for i, value := range valuesToAccumulate {
> ```

You're overthinking this. I'd rewrite the start of this loop body as:
```go
for i, value := range valuestoAccumulate {
   filterVal := parser.DNull
   if renderIdx, ok := n.filterToRenderIdxs[i]; ok {
      filterVal = values[renderIdx]
   }
   if filterVal == parser.DNull || filterVal == parser.DBoolFalse {
       continue
   }
   ...
```
   

---

*[pkg/sql/logic_test.go, line 1002 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-KUm1_azUeek2Q5DhTSm:-K_9zjDlE4XS7PfIPNXW:b-ekqhk9) ([raw file](https://github.com/cockroachdb/cockroach/blob/c55c2e79286f9596bb6969bf7a193a5248af8204/pkg/sql/logic_test.go#L1002)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

correct.
@knz, I remember you touching and fixing related stuff recently. Does this ring a bell?
</blockquote></details>

Yes I fixed `rowsort` and made sure it works as expected: 0689da591e75d079e3e36b57d09a9228c79f07d0

I think you can now use it in your tests and remove this comment.

---

*[pkg/sql/parser/type_check.go, line 405 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_9yvbhFXdZ9mJukZGt:-K_9yvbiOKfMGH1zVO1J:b8mxq4) ([raw file](https://github.com/cockroachdb/cockroach/blob/922c218ebbfd70e5f8d2e761e704c6709612ac52/pkg/sql/parser/type_check.go#L405)):*
> ```Go
> 
> 	if expr.Filter != nil {
> 		tExpr, err := expr.Filter.TypeCheck(ctx, TypeBool)
> ```

Use `typeCheckAndRequireBoolean`.

---

*[pkg/sql/parser/type_check.go, line 427 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_9yW6Gh_Kj9x_mLPnY:-K_9yW6H63XjabKDQLFA:b-oe2osx) ([raw file](https://github.com/cockroachdb/cockroach/blob/922c218ebbfd70e5f8d2e761e704c6709612ac52/pkg/sql/parser/type_check.go#L427)):*
> ```Go
> 
> 		if expr.Filter != nil {
> 			return nil, fmt.Errorf("filter within a window function call is not yet implemented")
> ```

make this error a global variable. Capitalize "FILTER". 

---

*[pkg/sql/parser/type_check.go, line 440 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_9yeU6Q2hQ813tk176:-K_9yeU6Q2hQ813tk177:bp8earv) ([raw file](https://github.com/cockroachdb/cockroach/blob/922c218ebbfd70e5f8d2e761e704c6709612ac52/pkg/sql/parser/type_check.go#L440)):*
> ```Go
> 		if builtin.class != AggregateClass {
> 			// Same error message as Postgres.
> 			return nil, fmt.Errorf("FILTER specified, but %s is not an aggregate function", expr.Func)
> ```

I'd like to clarify with `... but %s() is not an aggregate function` (with parentheses)

---

*[pkg/sql/testdata/aggregate, line 808 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-KUmu_GNJQ7FsA_9AfZo:-K_9zPaQD8dPFIN9VSYn:b-4cz8r1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c55c2e79286f9596bb6969bf7a193a5248af8204/pkg/sql/testdata/aggregate#L808)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

well I've added an `EXPLAIN(TYPES)` and I think we're seeing what we want, but someone double-checking would be great
</blockquote></details>

I think it's ok.

---

*[pkg/sql/testdata/aggregate, line 949 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_9z5XnCnyV1asievfR:-K_9z5XodehKXQKzR5l-:blumeva) ([raw file](https://github.com/cockroachdb/cockroach/blob/922c218ebbfd70e5f8d2e761e704c6709612ac52/pkg/sql/testdata/aggregate#L949)):*
> ```
> 
> 
> # #########################
> ```

We don't do test sections with this format. 
Either you think this is something we should generally do, then start by adding these section delimiters everywhere in this file where it makes sense (I see at least 3-4 sections already), or it is not and then simply add a single-line comment before the relevant tests you want to explain.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
It's a Christmas miracle! I feared we'd never hear about this change again.

I don't have any new remarks outside of what @knz said.

---

Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/sql/group.go, line 208 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_8BgnVBwGFh7i8ehFa:-K_AGgusgzqum8nQr9yr:b2teadv) ([raw file](https://github.com/cockroachdb/cockroach/blob/428f525528d39721afede723905ad7e14008cae7/pkg/sql/group.go#L208)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This comment is obviously garbage since the argument is false. I do not know why I forgot to remove it when I committed 9450059ada3d31fe73f0fe4573f4b39b4cbb9d79. 
FWIW I just made a small experiment by setting the argument to `true` and it breaks badly. I do not know yet why it breaks, but that is why the argument is currently `false` and probably why the comment was written (I probably wrote the comment first, thinking it could be `true`, then tweaked the argument later without realizing the comment was stale.)

Note that yes, we'll want to merge renders eventually. This is definitely on my radar still (here and in other places). However we need top clean up some other code before we can solve this issue.
</blockquote></details>

I'm pretty sure it breaks because of the `valuesToAccumulate/valuesToGroupBy` split in `groupNode.Next` (which inspired us to do the same thing with `windowNode`).

---

*[pkg/sql/group.go, line 233 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_A1sKrC475UDrxdX_O:-K_AH7AHfRlO70122F43:b-1ql25g) ([raw file](https://github.com/cockroachdb/cockroach/blob/922c218ebbfd70e5f8d2e761e704c6709612ac52/pkg/sql/group.go#L233)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

In this specific case I think you can request the merge with `true` and it should work, because you track the mapping manually in `filterToRenderIdxs`. Have you tried? Ideally with this if you have `select count(*) filter(where x+1=3), max(x) filter(where x+1=3) from ...` you should see `x+1=3` rendered just once (use EXPLAIN(EXPRS) to check).
</blockquote></details>

Yes because you access `values[renderIdx]` directly below, merging them should work.

---

*[pkg/sql/parser/type_check.go, line 405 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_9yvbhFXdZ9mJukZGt:-K_AHHW-1u46gTV170qm:b32v57) ([raw file](https://github.com/cockroachdb/cockroach/blob/922c218ebbfd70e5f8d2e761e704c6709612ac52/pkg/sql/parser/type_check.go#L405)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Use `typeCheckAndRequireBoolean`.
</blockquote></details>

+1

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-:-K_AGHh-86F2GPSZbVxW:b-dz5q5j)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Can you add a check in distsql physical planner CheckSupport to make sure we don't try to run these with distsql? And run LogicTest with COCKROACH_DISTSQL_MODE=on to make sure the new tests pass (note that `/views` will fail, there is an issue about it). Thanks!

andreimatei commented:
![giphy.gif](https://files.reviewableusercontent.io/images/reviewable/377201/fJ7Q3qLhSAiv-FLryDDA7CC6/giphy.gif)

@raduberinde I had rememberd last night that I need to do that too, done.
It's interesting what happens if I try to run with distsql: the filter is not simply ignored, but it seems to cause one of the stages of aggregation to group by the wrong combination of columns. Probably some render indexes gone wrong somewhere; any clue where I would look?

```
statement ok
CREATE TABLE filter_test (
  k INT,
  v INT,
  mark BOOL
)

statement OK
INSERT INTO filter_test VALUES
(1, 2, false),
(3, 4, true),
(5, NULL, true),
(6, 2, true),
(7, 2, true),
(8, 4, true),
(NULL, 4, true)


# FILTER should eliminate some results.
query II rowsort
SELECT v, COUNT(*) FILTER (WHERE k > 5) FROM filter_test GROUP BY v
----
2 2
4 1
NULL 0

!!! notice how we don't get unique "v"s.

    		expected:
    		    2 2
    		    4 1
    		    NULL 0

    		but found (query options: "rowsort" -> ignore the following ordering of rows) :
    		    NULL  1
    		    2     2
    		    4     1
    		    4     1
    		    2     1
    		    4     1
```

---

Review status: 5 of 13 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[pkg/sql/group.go, line 208 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_8BgnVBwGFh7i8ehFa:-K_AuRf6ibFfY88od9cK:b91cdpt) ([raw file](https://github.com/cockroachdb/cockroach/blob/428f525528d39721afede723905ad7e14008cae7/pkg/sql/group.go#L208)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'm pretty sure it breaks because of the `valuesToAccumulate/valuesToGroupBy` split in `groupNode.Next` (which inspired us to do the same thing with `windowNode`).
</blockquote></details>

Thanks. Added a TODO; I might be inclined to fix it.
The comment was always because you can have `SELECT k GROUP BY k`, right? Meaning renders don't need to have aggregation functions in them necessarily.

---

*[pkg/sql/group.go, line 233 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_A1sKrC475UDrxdX_O:-K_AuVprY6cgUm4xwgVv:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/922c218ebbfd70e5f8d2e761e704c6709612ac52/pkg/sql/group.go#L233)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Yes because you access `values[renderIdx]` directly below, merging them should work.
</blockquote></details>

done

---

*[pkg/sql/group.go, line 398 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_A2kPX6Q3Ym9eFVmod:-K_AoAGF3umuZ5bwHv7v:b-co3oa2) ([raw file](https://github.com/cockroachdb/cockroach/blob/922c218ebbfd70e5f8d2e761e704c6709612ac52/pkg/sql/group.go#L398)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

You're overthinking this. I'd rewrite the start of this loop body as:
```go
for i, value := range valuestoAccumulate {
   filterVal := parser.DNull
   if renderIdx, ok := n.filterToRenderIdxs[i]; ok {
      filterVal = values[renderIdx]
   }
   if filterVal == parser.DNull || filterVal == parser.DBoolFalse {
       continue
   }
   ...
```
   

</blockquote></details>

done, except I think you meant `DBoolTrue` in the beginning and I'm just comparing to `DBoolTrue` at the end.
I've also changed the code in `sqlbase.RunFilter` to just compare to `DBoolTrue`.

---

*[pkg/sql/logic_test.go, line 1002 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-KUm1_azUeek2Q5DhTSm:-K_AljBFVVuT8ttoe10g:bed9qw4) ([raw file](https://github.com/cockroachdb/cockroach/blob/c55c2e79286f9596bb6969bf7a193a5248af8204/pkg/sql/logic_test.go#L1002)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Yes I fixed `rowsort` and made sure it works as expected: 0689da591e75d079e3e36b57d09a9228c79f07d0

I think you can now use it in your tests and remove this comment.
</blockquote></details>

the issue still stands, but I've removed the TODO and replaced it with a comment, because I'm not sure what the ideal error would be, and also improved the error message a bit.

---

*[pkg/sql/parser/type_check.go, line 405 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_9yvbhFXdZ9mJukZGt:-K_Am_iiRFKLcxZtvHzu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/922c218ebbfd70e5f8d2e761e704c6709612ac52/pkg/sql/parser/type_check.go#L405)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

+1
</blockquote></details>

Done.

---

*[pkg/sql/parser/type_check.go, line 427 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_9yW6Gh_Kj9x_mLPnY:-K_ApdxsTqQ5hYP1C-JM:bs58suc) ([raw file](https://github.com/cockroachdb/cockroach/blob/922c218ebbfd70e5f8d2e761e704c6709612ac52/pkg/sql/parser/type_check.go#L427)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

make this error a global variable. Capitalize "FILTER". 
</blockquote></details>

capitalized, but I don't think this deserves a global... There's no reason for one, and other similar errors in this files don't have it.

---

*[pkg/sql/parser/type_check.go, line 440 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_9yeU6Q2hQ813tk176:-K_Aq8Gn-u-awBIOhGu-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/922c218ebbfd70e5f8d2e761e704c6709612ac52/pkg/sql/parser/type_check.go#L440)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I'd like to clarify with `... but %s() is not an aggregate function` (with parentheses)
</blockquote></details>

Done.

---

*[pkg/sql/testdata/aggregate, line 949 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_9z5XnCnyV1asievfR:-K_AqALQuvqKFrwJ2PuJ:b-ygh9ca) ([raw file](https://github.com/cockroachdb/cockroach/blob/922c218ebbfd70e5f8d2e761e704c6709612ac52/pkg/sql/testdata/aggregate#L949)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

We don't do test sections with this format. 
Either you think this is something we should generally do, then start by adding these section delimiters everywhere in this file where it makes sense (I see at least 3-4 sections already), or it is not and then simply add a single-line comment before the relevant tests you want to explain.
</blockquote></details>

I do think this is something we should be doing, and I think we can do it incrementally. Or not. But consistency doesn't seem important.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-:-K_B-z-H0ELcStcnOLN7:bxcu3la)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 8 of 8 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/sql/logic_test.go, line 1029 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_BG4QZTrqaKrfLJ2ct:-K_BG4QZTrqaKrfLJ2cu:b-toxeve) ([raw file](https://github.com/cockroachdb/cockroach/blob/25125fdace66ba71f5e00d619f5bfacb2cf8a628/pkg/sql/logic_test.go#L1029)):*
> ```Go
> 			sortMsg = " -> ignore the following ordering of rows"
> 		}
> 		fmt.Fprintf(&buf, "but found (query options: %q%s) :\n", query.rawOpts, sortMsg)
> ```

nit: put this in a separate commit. Modify the existing tests that check for expected errors, if any, in that separate commit.

---

*[pkg/sql/parser/type_check.go, line 427 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_9yW6Gh_Kj9x_mLPnY:-K_BGJh9L1TMwYDLk8qM:bblvf4n) ([raw file](https://github.com/cockroachdb/cockroach/blob/922c218ebbfd70e5f8d2e761e704c6709612ac52/pkg/sql/parser/type_check.go#L427)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

capitalized, but I don't think this deserves a global... There's no reason for one, and other similar errors in this files don't have it.
</blockquote></details>

I think the rule was if there is no parameter in the error message then we should call the error allocator just once.

---

*[pkg/sql/testdata/aggregate, line 949 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_9z5XnCnyV1asievfR:-K_BGX-wWcZlrp3yrL9-:bj6delz) ([raw file](https://github.com/cockroachdb/cockroach/blob/922c218ebbfd70e5f8d2e761e704c6709612ac52/pkg/sql/testdata/aggregate#L949)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I do think this is something we should be doing, and I think we can do it incrementally. Or not. But consistency doesn't seem important.
</blockquote></details>

Huh wait what. Consistency *is* important.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/sql/logic_test.go, line 1029 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_BG4QZTrqaKrfLJ2ct:-K_Bd4icUCJXY3SbEZ2H:batsc3k) ([raw file](https://github.com/cockroachdb/cockroach/blob/25125fdace66ba71f5e00d619f5bfacb2cf8a628/pkg/sql/logic_test.go#L1029)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

nit: put this in a separate commit. Modify the existing tests that check for expected errors, if any, in that separate commit.
</blockquote></details>

but this doesn't affect tests expecting a particular error msg... I'll leave it here.

---

*[pkg/sql/parser/type_check.go, line 427 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_9yW6Gh_Kj9x_mLPnY:-K_BdKTLH2H3dm60LJUb:b-gc9qli) ([raw file](https://github.com/cockroachdb/cockroach/blob/922c218ebbfd70e5f8d2e761e704c6709612ac52/pkg/sql/parser/type_check.go#L427)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I think the rule was if there is no parameter in the error message then we should call the error allocator just once.
</blockquote></details>

but how many times we call the allocator for this really doesn't matter :)
If anything, it's a good thing that we're saving a few bytes of heap space whenever this error is not used (which is always) :P

---

*[pkg/sql/testdata/aggregate, line 949 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_9z5XnCnyV1asievfR:-K_Bd_ep-Nh01WIiRdd4:bdcufrk) ([raw file](https://github.com/cockroachdb/cockroach/blob/922c218ebbfd70e5f8d2e761e704c6709612ac52/pkg/sql/testdata/aggregate#L949)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Huh wait what. Consistency *is* important.
</blockquote></details>

not consistency for this thing. It's useful even as a one off.

---

*[pkg/sql/testdata/aggregate, line 1006 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_Bcg0mPxSRQgi1xXRB:-K_Bcg0mPxSRQgi1xXRC:b72msif) ([raw file](https://github.com/cockroachdb/cockroach/blob/25125fdace66ba71f5e00d619f5bfacb2cf8a628/pkg/sql/testdata/aggregate#L1006)):*
> ```
> 0                 render 1     max(k > 5) FILTER (WHERE k > 5)
> 1  render/filter
> 1                 render 0     *
> ```

rando question: what does this "render *" mean? I hope it doesn't mean we're rendering all the columns?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/sql/testdata/aggregate, line 949 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_9z5XnCnyV1asievfR:-K_BgoXH9pF1HWG6VYQM:b-z8szfd) ([raw file](https://github.com/cockroachdb/cockroach/blob/922c218ebbfd70e5f8d2e761e704c6709612ac52/pkg/sql/testdata/aggregate#L949)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

not consistency for this thing. It's useful even as a one off.
</blockquote></details>

I disagree; this abuse of the pound sign is disruptive to the reader, and this is a bad thing.

---

*[pkg/sql/testdata/aggregate, line 1006 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_Bcg0mPxSRQgi1xXRB:-K_BgyFEjA5FmTIh4hJH:b-ged6c5) ([raw file](https://github.com/cockroachdb/cockroach/blob/25125fdace66ba71f5e00d619f5bfacb2cf8a628/pkg/sql/testdata/aggregate#L1006)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

rando question: what does this "render *" mean? I hope it doesn't mean we're rendering all the columns?
</blockquote></details>

No it's doesn't mean that :)
This "*" is the result of pretty-printing the aggFuncHolder for `count(*)`, which merely pretty-prints the argument to the aggregate function, in this case a StarDatumInstance.

This is rather confusing, I agree, and if you could modify `func (a *aggregateFuncHolder) Format()` to indicate this is an aggregate func holder, it would be great. Like this:
```
func (a *aggregateFuncHolder) Format(buf *bytes.Buffer, f parser.FmtFlags) {
   buf.WriteString("[agg arg ")
   a.expr.Format(buf, f)
   buf.WriteByte(']')
}
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/sql/testdata/aggregate, line 949 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_9z5XnCnyV1asievfR:-K_Bl2kE8_RNIGYYfjgn:b-q5psa1) ([raw file](https://github.com/cockroachdb/cockroach/blob/922c218ebbfd70e5f8d2e761e704c6709612ac52/pkg/sql/testdata/aggregate#L949)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I disagree; this abuse of the pound sign is disruptive to the reader, and this is a bad thing.
</blockquote></details>

well don't you agree that it'd be nice to have sections in these files? And if you agree, isn't one section better than 0 sections?
What you call "disruptive" I call the only line worth reading in most cases when you're scrolling through this file.

---

*[pkg/sql/testdata/aggregate, line 1006 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_Bcg0mPxSRQgi1xXRB:-K_Bn1XkjogwossuKagF:b5ruuge) ([raw file](https://github.com/cockroachdb/cockroach/blob/25125fdace66ba71f5e00d619f5bfacb2cf8a628/pkg/sql/testdata/aggregate#L1006)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

No it's doesn't mean that :)
This "*" is the result of pretty-printing the aggFuncHolder for `count(*)`, which merely pretty-prints the argument to the aggregate function, in this case a StarDatumInstance.

This is rather confusing, I agree, and if you could modify `func (a *aggregateFuncHolder) Format()` to indicate this is an aggregate func holder, it would be great. Like this:
```
func (a *aggregateFuncHolder) Format(buf *bytes.Buffer, f parser.FmtFlags) {
   buf.WriteString("[agg arg ")
   a.expr.Format(buf, f)
   buf.WriteByte(']')
}
```
</blockquote></details>

but changing `aggregateFuncHolder.Format()` wouldn't change this line of output (notice this is under the "render/filter" stage). It would change the "render" lines from the "group" stage, which are fine.
We need to somehow change the formatting of the expression that we pass to the lower `selectNode`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/sql/testdata/aggregate, line 949 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_9z5XnCnyV1asievfR:-K_BoRRTydEG1xiouf8U:bhz4cmf) ([raw file](https://github.com/cockroachdb/cockroach/blob/922c218ebbfd70e5f8d2e761e704c6709612ac52/pkg/sql/testdata/aggregate#L949)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

well don't you agree that it'd be nice to have sections in these files? And if you agree, isn't one section better than 0 sections?
What you call "disruptive" I call the only line worth reading in most cases when you're scrolling through this file.
</blockquote></details>

Dude it's either many pound signs *and* multiple section headers in this file, or not at all. Both the many pound signs and the inconsistency are grating.

---

*[pkg/sql/testdata/aggregate, line 1006 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_Bcg0mPxSRQgi1xXRB:-K_BoGUsLHKC0TxuuUOc:bjyqjht) ([raw file](https://github.com/cockroachdb/cockroach/blob/25125fdace66ba71f5e00d619f5bfacb2cf8a628/pkg/sql/testdata/aggregate#L1006)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

but changing `aggregateFuncHolder.Format()` wouldn't change this line of output (notice this is under the "render/filter" stage). It would change the "render" lines from the "group" stage, which are fine.
We need to somehow change the formatting of the expression that we pass to the lower `selectNode`.
</blockquote></details>

I think you should try it and see for yourself.

It's not possible that a naked StarDatumInstance lands in the render list, otherwise Eval would panic. If you can see it, it's necessarily enclosed in an aggFuncHolder.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 11 of 13 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/sql/testdata/aggregate, line 949 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_9z5XnCnyV1asievfR:-K_BqOqAIqp-ulAtpuez:b-gvw635) ([raw file](https://github.com/cockroachdb/cockroach/blob/922c218ebbfd70e5f8d2e761e704c6709612ac52/pkg/sql/testdata/aggregate#L949)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Dude it's either many pound signs *and* multiple section headers in this file, or not at all. Both the many pound signs and the inconsistency are grating.
</blockquote></details>

ok, removed the many pounds. But if we had other sections, wouldn't it be even more disruptive? :)

---

*[pkg/sql/testdata/aggregate, line 1006 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_Bcg0mPxSRQgi1xXRB:-K_BpTe1qTHcQ0d4V_yz:b-rffwpb) ([raw file](https://github.com/cockroachdb/cockroach/blob/25125fdace66ba71f5e00d619f5bfacb2cf8a628/pkg/sql/testdata/aggregate#L1006)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I think you should try it and see for yourself.

It's not possible that a naked StarDatumInstance lands in the render list, otherwise Eval would panic. If you can see it, it's necessarily enclosed in an aggFuncHolder.
</blockquote></details>

I did try:

With the change, on some query:
```
  		    0  group
    		    0                 aggregate 0  count(k)
    		    0                 render 0     count(k)
    		    1  render/filter
    		    1                 render 0     k
    		    2  scan
    		    2                 table        kv@primary

    		but found (query options: "") :
    		    0  group
    		    0                 aggregate 0  count(k)
    		    0                 render 0     [agg arg "count(k)"]
    		    1  render/filter
    		    1                 render 0     k
    		    2  scan
    		    2                 table        kv@primary
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 11 of 13 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/sql/testdata/aggregate, line 1006 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_Bcg0mPxSRQgi1xXRB:-K_Br_VtsrcehYVYmhgR:blcexpw) ([raw file](https://github.com/cockroachdb/cockroach/blob/25125fdace66ba71f5e00d619f5bfacb2cf8a628/pkg/sql/testdata/aggregate#L1006)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I did try:

With the change, on some query:
```
  		    0  group
    		    0                 aggregate 0  count(k)
    		    0                 render 0     count(k)
    		    1  render/filter
    		    1                 render 0     k
    		    2  scan
    		    2                 table        kv@primary

    		but found (query options: "") :
    		    0  group
    		    0                 aggregate 0  count(k)
    		    0                 render 0     [agg arg "count(k)"]
    		    1  render/filter
    		    1                 render 0     k
    		    2  scan
    		    2                 table        kv@primary
```
</blockquote></details>

I pasted an example without star above, but here's one with it:
```
root@:26257> EXPLAIN (exprs) SELECT 1+COUNT(*) FILTER (WHERE k > 5), MAX(k>5) FILTER (WHERE k > 5) FROM filter_test GROUP BY v;
+-------+---------------+-------------+-----------------------------------------------------------+
| Level |     Type      |    Field    |                        Description                        |
+-------+---------------+-------------+-----------------------------------------------------------+
|     0 | group         |             |                                                           |
|     0 |               | aggregate 0 |                                                         1 |
|     0 |               | aggregate 1 | count(*) FILTER (WHERE k > 5)                             |
|     0 |               | aggregate 2 | max(k > 5) FILTER (WHERE k > 5)                           |
|     0 |               | render 0    | [agg arg "1"] + [agg arg "count(*) FILTER (WHERE k > 5)"] |
|     0 |               | render 1    | [agg arg "max(k > 5) FILTER (WHERE k > 5)"]               |
|     1 | render/filter |             |                                                           |
|     1 |               | render 0    |                                                         1 |
|     1 |               | render 1    | *                                                         |
|     1 |               | render 2    | k > 5                                                     |
|     1 |               | render 3    | v                                                         |
|     2 | scan          |             |                                                           |
|     2 |               | table       | filter_test@primary                                       |
+-------+---------------+-------------+-----------------------------------------------------------+
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146)*
<!-- Sent from Reviewable.io -->

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> (modulo the distsql issue)

---

Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/sql/group.go, line 208 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_8BgnVBwGFh7i8ehFa:-K_C5LAnZt_n5cJ_gTOz:b-lwj2cr) ([raw file](https://github.com/cockroachdb/cockroach/blob/428f525528d39721afede723905ad7e14008cae7/pkg/sql/group.go#L208)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Thanks. Added a TODO; I might be inclined to fix it.
The comment was always because you can have `SELECT k GROUP BY k`, right? Meaning renders don't need to have aggregation functions in them necessarily.
</blockquote></details>

This was already filed as #10919.

---

*[pkg/sql/testdata/aggregate, line 949 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_9z5XnCnyV1asievfR:-K_C5QaXSE10BjUGavU-:b-z2f5q8) ([raw file](https://github.com/cockroachdb/cockroach/blob/922c218ebbfd70e5f8d2e761e704c6709612ac52/pkg/sql/testdata/aggregate#L949)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

ok, removed the many pounds. But if we had other sections, wouldn't it be even more disruptive? :)
</blockquote></details>

I think that if there were multiple such "section headers" this particular disruptiveness would be compensated by the improvement in overall clarify of the test file. :smile_cat: 

---

*[pkg/sql/testdata/aggregate, line 1006 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_Bcg0mPxSRQgi1xXRB:-K_C5eoxsm8GlVXst4HH:b-u3iexh) ([raw file](https://github.com/cockroachdb/cockroach/blob/25125fdace66ba71f5e00d619f5bfacb2cf8a628/pkg/sql/testdata/aggregate#L1006)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I pasted an example without star above, but here's one with it:
```
root@:26257> EXPLAIN (exprs) SELECT 1+COUNT(*) FILTER (WHERE k > 5), MAX(k>5) FILTER (WHERE k > 5) FROM filter_test GROUP BY v;
+-------+---------------+-------------+-----------------------------------------------------------+
| Level |     Type      |    Field    |                        Description                        |
+-------+---------------+-------------+-----------------------------------------------------------+
|     0 | group         |             |                                                           |
|     0 |               | aggregate 0 |                                                         1 |
|     0 |               | aggregate 1 | count(*) FILTER (WHERE k > 5)                             |
|     0 |               | aggregate 2 | max(k > 5) FILTER (WHERE k > 5)                           |
|     0 |               | render 0    | [agg arg "1"] + [agg arg "count(*) FILTER (WHERE k > 5)"] |
|     0 |               | render 1    | [agg arg "max(k > 5) FILTER (WHERE k > 5)"]               |
|     1 | render/filter |             |                                                           |
|     1 |               | render 0    |                                                         1 |
|     1 |               | render 1    | *                                                         |
|     1 |               | render 2    | k > 5                                                     |
|     1 |               | render 3    | v                                                         |
|     2 | scan          |             |                                                           |
|     2 |               | table       | filter_test@primary                                       |
+-------+---------------+-------------+-----------------------------------------------------------+
```
</blockquote></details>

Woah, this is ... unfortunate. Mind filing an issue - assign to me or you - with title "sql: StarDatumInstance improperly added to the render list" and this EXPLAIN output in the explanation.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-:-K_C5yCCc_HD81_qxVZG:b-3fme3c)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
distsql planner change LGTM, thanks!

I would start by looking at the plan: use COCKROACH_DISTSQL_LOG_PLAN=1 and then paste the logged JSON into http://raduberinde.github.io  The expressions in the aggregator box should show what it's actually doing. I don't know if it's worth debugging except as part of actually making these filters work with distsql (which we can do later). 

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-:-K_CjU5rJQU88bVmuXW6:b-s1xklu)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/testdata/aggregate, line 1006 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146#-K_Bcg0mPxSRQgi1xXRB:-K_FwttOk4pei-172zww:bwiuyct) ([raw file](https://github.com/cockroachdb/cockroach/blob/25125fdace66ba71f5e00d619f5bfacb2cf8a628/pkg/sql/testdata/aggregate#L1006)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Woah, this is ... unfortunate. Mind filing an issue - assign to me or you - with title "sql: StarDatumInstance improperly added to the render list" and this EXPLAIN output in the explanation.
</blockquote></details>

filed #12615 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10146)*
<!-- Sent from Reviewable.io -->
